bin/kafka-run-class.sh

```diff
@@ -62,80 +62,80 @@ do
   if [ -z "$CLASSPATH" ] ; then
     CLASSPATH="$dir/*"
   else
-    CLASSPATH="$CLASSPATH:$dir/*"
+    CLASSPATH="$dir/*":$CLASSPATH
```

Comment:
We'd need to ensure the ordering works ok here as well. I'm not sure people will expect things to be prepended. If you've already set a `CLASSPATH` to override some jar, you won't get the same behavior anymore.

Also, do we know why have an empty classpath segment at the end doesn't cause the problem but at the beginning it works? And is it even a valid assumption across different JDKs?

Reply:
After doing some digging it looks like the reason appending CLASSPATH instead of prepending it works is due to the use of the [String.split()](https://docs.oracle.com/javase/7/docs/api/java/lang/String.html#split(java.lang.String)) method on the classpath environment variable in the Reflections API. An empty string would not be found at the end of the resulting array for a call to something like `"test:classpath:".split(":")` but would be at the beginning for something like `":test:classpath".split(":")`. This functionality, being documented in the String API, seems like it'd be the same across every (valid) JDK implementation.

However, I think @ewencp's point about wanting to allow jar overriding still stands. A possible fix that keeps that option for the user would be to add the line `CLASSPATH=${CLASSPATH#:}` to the run-class script before before the JVM is started (and also before the necessary conversion to a Windows-compliant classpath in the event the user is running Cygwin) and after all the necessary directories/jars have been added to it. I'd be happy to create a separate PR for that fix if you don't have time, LMK!

Reply:
New PR is up at https://github.com/apache/kafka/pull/4406, as requested by @rnpridgeon offline.